#  Decision Trees Regression & Random Forest Method

We are going to be using decision tree regression and random forest method to predict the uber fare prices. We are going to be using the universal features from the models and creating a decision tree out of it. As part of the decision tree algorithm its going to select the important features itself and is going to make a tree depending on the max_dept we input. At the end we are going to be comparing the predicted values with the actual values and see the MSE difference. 

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error
import matplotlib.pyplot as plt

In [3]:
df = pd.read_csv('../data/data_feature_engineering.csv')

FileNotFoundError: [Errno 2] No such file or directory: '../data/data_feature_engineering.csv'

In [ ]:
print(df.head())

                                     id   timestamp  day_of_week  day  time  \
0  009e9c53-074d-43cf-aef2-0fbc7a47ed3d  1543615981            0   30  1189   
1  23f145da-f0c1-4d1f-a184-496bc003a7db  1544698211            4   13   562   
2  357559cb-8c58-4278-a41a-e33b2e0997a3  1544728504            4   13  1034   
3  50ef1165-9d23-416c-a65c-18906207b295  1545004511            3   16  1279   
4  91c4861c-1780-42b0-bca1-bbd64a422cc3  1544748008            0   14    35   

   datetime  hour  source  destination  name  ...  precipIntensityMax  \
0      8565    22       6           11     4  ...              0.0003   
1     17198    10       6           11     0  ...              0.0001   
2     18045    19       6           11     3  ...              0.0001   
3     25775    23       6           11     5  ...              0.1252   
4     18593     0       6           11     1  ...              0.0001   

   uvIndexTime  temperatureMin  temperatureMinTime  temperatureMax  \
0   1543593600  

In [67]:
feature_columns = ['name','distance', 'source','precipIntensity','humidity', 'temperatureHigh' , 'apparentTemperatureHigh', 'uvIndex','precipIntensityMax','temperatureMax','apparentTemperatureMax']
X = df[feature_columns]

In [68]:
y = df['price'] 

In [69]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [70]:
model1 = DecisionTreeRegressor(max_depth=10, random_state=42)
model1.fit(X_train, y_train)

DecisionTreeRegressor(max_depth=10, random_state=42)

We changed the Max_depth to a range of values from 2 to 100 and we based our final result on the lowest MSE keeping overfitting and underfitting in mind. In this case the change in max dept for a high number such as 100 lead to a higher MSE value.

In [59]:
from sklearn.tree import DecisionTreeRegressor, export_text


In [60]:
model1_text = export_text(model1, feature_names=feature_columns)
print(model1_text)

|--- name <= 1.50
|   |--- name <= 0.50
|   |   |--- distance <= 2.20
|   |   |   |--- distance <= 1.64
|   |   |   |   |--- distance <= 1.35
|   |   |   |   |   |--- distance <= 1.11
|   |   |   |   |   |   |--- source <= 3.50
|   |   |   |   |   |   |   |--- apparentTemperatureMax <= 50.30
|   |   |   |   |   |   |   |   |--- distance <= 1.09
|   |   |   |   |   |   |   |   |   |--- precipIntensity <= 0.12
|   |   |   |   |   |   |   |   |   |   |--- value: [15.28]
|   |   |   |   |   |   |   |   |   |--- precipIntensity >  0.12
|   |   |   |   |   |   |   |   |   |   |--- value: [16.67]
|   |   |   |   |   |   |   |   |--- distance >  1.09
|   |   |   |   |   |   |   |   |   |--- precipIntensityMax <= 0.13
|   |   |   |   |   |   |   |   |   |   |--- value: [15.20]
|   |   |   |   |   |   |   |   |   |--- precipIntensityMax >  0.13
|   |   |   |   |   |   |   |   |   |   |--- value: [14.83]
|   |   |   |   |   |   |   |--- apparentTemperatureMax >  50.30
|   |   |   |   |   |   |   

In [61]:
y_pred1 = model1.predict(X_test)

In [62]:
print("Actual vs. Predicted values:\n")
for actual, predicted in zip(y_test[:10], y_pred1[:10]):
    print(f"Actual: {actual:.2f} - Predicted: {predicted:.2f}")

Actual vs. Predicted values:

Actual: 5.50 - Predicted: 6.55
Actual: 8.00 - Predicted: 7.24
Actual: 18.50 - Predicted: 18.03
Actual: 20.50 - Predicted: 19.25
Actual: 23.50 - Predicted: 22.31
Actual: 9.50 - Predicted: 8.58
Actual: 31.50 - Predicted: 31.06
Actual: 9.50 - Predicted: 10.69
Actual: 8.50 - Predicted: 8.05
Actual: 9.50 - Predicted: 9.90


In [63]:
print("\nModel Mean Squared Error:", mean_squared_error(y_test, y_pred1))



Model Mean Squared Error: 3.6982944198879424


This is the error we are getting between the predicted values and the original values. This error is low suggesting how decision tree model could potantially be the best machine learning model. 

In [14]:
print(model1.feature_importances_)

[8.51115518e-01 1.44312230e-01 3.47095049e-03 1.60319889e-04
 2.20248880e-04 1.26250980e-04 9.84395687e-05 5.74504464e-05
 1.57038120e-04 1.18638969e-04 1.62913782e-04]


This represent the importance of each feature. We can clearly see how name and distance are one of the most important features to determine the fare rate for a decision tree model. An interesting think to note is as we increase the max_dept the importance of distance decreases and is distributed to other features.

# Random Forest Regressor 

In [18]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

In [19]:
feature_columns = ['name','distance', 'source','precipIntensity','humidity', 'temperatureHigh' , 'apparentTemperatureHigh', 'uvIndex','precipIntensityMax','temperatureMax','apparentTemperatureMax']
X = df[feature_columns]
y = df['price'] 

In [20]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)


In [21]:
model2 = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=42)

This is the random forest regressor with the same max_dept as all the models above so that the test is fair. 

In [22]:
model2.fit(X_train, y_train)

RandomForestRegressor(max_depth=10, random_state=42)

In [23]:
y_pred2 = model2.predict(X_test)

In [24]:
print("Actual vs. Predicted values:\n")
for actual, predicted in zip(y_test[:10], y_pred2[:10]):
    print(f"Actual: {actual:.2f} - Predicted: {predicted:.2f}")

Actual vs. Predicted values:

Actual: 10.50 - Predicted: 11.53
Actual: 10.50 - Predicted: 10.89
Actual: 8.00 - Predicted: 8.34
Actual: 17.00 - Predicted: 18.81
Actual: 17.00 - Predicted: 16.78
Actual: 7.50 - Predicted: 7.25
Actual: 14.50 - Predicted: 14.09
Actual: 14.50 - Predicted: 10.05
Actual: 11.50 - Predicted: 10.41
Actual: 7.50 - Predicted: 7.18


In [25]:
print("\nModel Mean Squared Error:", mean_squared_error(y_test, y_pred2))


Model Mean Squared Error: 3.533933453651921


This is the best MSE we have gotten but marginally better than the one for decision tree. 

Through the application of Decision Tree and Random Forest regressors, we achieved notably low Mean Squared Error (MSE) values, some as low as 3.5, in predicting Uber prices using all available features. This success underscores the potential of these models in understanding the complexities of Uber's pricing dynamics. However, challenges persist, particularly in feature selection, as the current set may not fully capture all variables influencing dynamic pricing. The intricate nature of the pricing model, including non-linear relationships and potential overfitting, further complicates the predictive process. Future improvements could focus on refining feature selection, model complexity, and addressing potential underfitting or overfitting to enhance prediction accuracy.